internal/ls/utilities.go

```diff
@@ -0,0 +1,376 @@
+			return node.Parent
+		}
+		return nil
+	case ast.KindLiteralType:
+		if !ast.IsStringLiteral(node) {
+			return nil
+		}
+		if ast.IsImportTypeNode(node.Parent.Parent) {
+			return node.Parent.Parent
+		}
+		return nil
+	}
+	return nil
+}
+
+// !!!
+func isInComment(file *ast.SourceFile, position int, tokenAtPosition *ast.Node) *ast.CommentRange {
+	return nil
+}
+
+// !!!
+// Replaces last(node.getChildren(sourceFile))
+func getLastChild(node *ast.Node, sourceFile *ast.SourceFile) *ast.Node {
+	return nil
+}
+
+func getLastToken(node *ast.Node, sourceFile *ast.SourceFile) *ast.Node {
+	if node == nil {
+		return nil
+	}
+
+	assertHasRealPosition(node)
+
+	lastChild := getLastChild(node, sourceFile)
+	if lastChild == nil {
+		return nil
+	}
+
+	if lastChild.Kind < ast.KindFirstNode {
+		return lastChild
+	} else {
+		return getLastToken(lastChild, sourceFile)
+	}
+}
+
+// !!!
+func getFirstToken(node *ast.Node, sourceFile *ast.SourceFile) *ast.Node {
+	return nil
+}
+
+func assertHasRealPosition(node *ast.Node) {
+	if ast.PositionIsSynthesized(node.Pos()) || ast.PositionIsSynthesized(node.End()) {
+		panic("Node must have a real position for this operation.")
+	}
+}
+
+// !!!
+func findChildOfKind(node *ast.Node, kind ast.Kind, sourceFile *ast.SourceFile) *ast.Node {
+	return nil
+}
+
+// !!!
+type PossibleTypeArgumentInfo struct {
+	called         *ast.IdentifierNode
+	nTypeArguments int
+}
+
+// !!!
+func getPossibleTypeArgumentsInfo(tokenIn *ast.Node, sourceFile *ast.SourceFile) *PossibleTypeArgumentInfo {
+	return nil
+}
+
+// !!!
+func getPossibleGenericSignatures(called *ast.Expression, typeArgumentCount int, checker *checker.Checker) []*checker.Signature {
+	return nil
+}
+
+func isInRightSideOfInternalImportEqualsDeclaration(node *ast.Node) bool {
+	for node.Parent.Kind == ast.KindQualifiedName {
+		node = node.Parent
+	}
+
+	return ast.IsInternalModuleImportEqualsDeclaration(node.Parent) && node.Parent.AsImportEqualsDeclaration().ModuleReference == node
+}
+
+func (l *LanguageService) createLspRangeFromNode(node *ast.Node, file *ast.SourceFile) *lsproto.Range {
+	return l.createLspRangeFromBounds(node.Pos(), node.End(), file)
+}
+
+func (l *LanguageService) createLspRangeFromBounds(start, end int, file *ast.SourceFile) *lsproto.Range {
+	lspRange, err := l.converters.ToLSPRange(file.FileName(), core.NewTextRange(start, end))
+	if err != nil {
+		panic(err)
+	}
+	return &lspRange
+}
+
+func quote(file *ast.SourceFile, preferences *UserPreferences, text string) string {
+	// Editors can pass in undefined or empty string - we want to infer the preference in those cases.
+	quotePreference := getQuotePreference(file, preferences)
+	quoted, _ := core.StringifyJson(text, "" /*prefix*/, "" /*indent*/)
+	if quotePreference == quotePreferenceSingle {
+		strings.ReplaceAll(strings.ReplaceAll(core.StripQuotes(quoted), "'", `\'`), `\"`, `"`)
```

Comment:
For this, I would suggest using `strings.NewReplacer` in a global var; it's faster.